PWGHF/D2H/Tasks/taskXicToXiPiPi.cxx

```diff
@@ -432,8 +430,69 @@ struct HfTaskXicToXiPiPi {
+  {
+    fillHistograms(candidates);
+
+    // fill KFParticle specific histograms
+    for (const auto& candidate : candidates) {
+      if (yCandRecoMax >= 0. && std::abs(candidate.y(o2::constants::physics::MassXiCPlus)) > yCandRecoMax) {
+        continue;
+      }
+      registry.fill(HIST("hChi2topoToPV"), candidate.chi2TopoXicPlusToPV(), candidate.pt());
+      registry.fill(HIST("hChi2topoXiToXicPlus"), candidate.chi2TopoXiToXicPlus(), candidate.pt());
+      registry.fill(HIST("hChi2geoXi"), candidate.kfCascadeChi2(), candidate.pt());
+      registry.fill(HIST("hChi2geoLam"), candidate.kfV0Chi2(), candidate.pt());
+    }
+  }
+  PROCESS_SWITCH(HfTaskXicToXiPiPi, processWithKFParticle, "Process Run 3 with KFParticle", false);
+
+  /// MC analysis and fill histograms
+  void processMcWithDCAFitter(soa::Filtered<soa::Join<aod::HfCandXic, aod::HfSelXicToXiPiPi, aod::HfCandXicMcRec>> const& candidates,
+                              soa::Join<aod::McParticles, aod::HfCandXicMcGen> const& mcParticles,
+                              aod::TracksWMc const& tracksWMc)
+  {
+    fillHistogramsMC(candidates, mcParticles, tracksWMc);
+  }
+  PROCESS_SWITCH(HfTaskXicToXiPiPi, processMcWithDCAFitter, "Process MC", false);
+
+  /// MC analysis and fill histograms with KFParticle
+  void processMcWithKFParticle(soa::Filtered<soa::Join<aod::HfCandXic, aod::HfCandXicKF, aod::HfSelXicToXiPiPi, aod::HfCandXicMcRec>> const& candidates,
+                               soa::Join<aod::McParticles, aod::HfCandXicMcGen> const& mcParticles,
+                               aod::TracksWMc const& tracksWMc)
+  {
+    fillHistogramsMC(candidates, mcParticles, tracksWMc);
+
+    // fill KFParticle specific histograms
+    for (const auto& candidate : candidates) {
+      if (yCandRecoMax >= 0. && std::abs(candidate.y(o2::constants::physics::MassXiCPlus)) > yCandRecoMax) {
+        continue;
+      }
+
+      int flagMcMatchRecXic = std::abs(candidate.flagMcMatchRec());
+
+      if (TESTBIT(flagMcMatchRecXic, hf_cand_xic_to_xi_pi_pi::DecayType::XicToXiPiPi) || TESTBIT(flagMcMatchRecXic, hf_cand_xic_to_xi_pi_pi::DecayType::XicToXiResPiToXiPiPi)) {
+        registry.fill(HIST("hChi2topoToPVRecSig"), candidate.chi2TopoXicPlusToPV(), candidate.pt());
+        registry.fill(HIST("hChi2topoXiToXicPlusRecSig"), candidate.chi2TopoXiToXicPlus(), candidate.pt());
+        registry.fill(HIST("hChi2geoXiRecSig"), candidate.kfCascadeChi2(), candidate.pt());
+        registry.fill(HIST("hChi2geoLamRecSig"), candidate.kfV0Chi2(), candidate.pt());
+      } else {
+        registry.fill(HIST("hChi2topoToPVRecBg"), candidate.chi2TopoXicPlusToPV(), candidate.pt());
+        registry.fill(HIST("hChi2topoXiToXicPlusRecBg"), candidate.chi2TopoXiToXicPlus(), candidate.pt());
+        registry.fill(HIST("hChi2geoXiRecBg"), candidate.kfCascadeChi2(), candidate.pt());
+        registry.fill(HIST("hChi2geoLamRecBg"), candidate.kfV0Chi2(), candidate.pt());
+      }
```

Comment:
This can be included in `fillHistogramsMC` inside an `if constexpr (useKfParticle) {}` block, using `bool useKfParticle` as a template argument of `fillHistogramsMC`.